### PR TITLE
#165596316 Setup a mail class and integrate sendgrid

### PR DESCRIPTION
**Pull request by devPinheiro**
#### What does this PR do?
- It contains a feature for sending emails via a 3rd party mail service

#### Description of Task to be completed?
- add test for edge cases
- install sendgrid package
- update .env.sample with API key
- create helper mail class
- implement chainable methods within the mail class
- setup mail.config file for 3rd party mailing system

#### How should this be manually tested?
- After cloning the repo, cd into the root folder

- Install all necessary dependencies by running 

  ``npm install``

- Launch the node server with

  ``npm run dev``

- To run local tests using mocha
  
  ``npm run test``

#### Any background context you want to provide?
- You should have Nodejs installed

#### What are the relevant pivotal tracker stories?
- PT Story link - `[#165596316] (https://www.pivotaltracker.com/story/show/165596316)`

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -1,4 +1,5 @@
 DATABASE_CLIENT=
 DATABASE_URL=
 TEST_DATABASE_CLIENT=
-TEST_DATABASE_URL=
\ No newline at end of file
+TEST_DATABASE_URL=
+SENDGRID_API_KEY=
```
Comment (bahdcoder) [inline .env.sample]: Please can you add this to the readme.md file ? some sort of documentation so others can know how to use. You could also share your API key on slack :)

File: test/index.spec.js [hunk 2]
```diff
@@ -1,13 +1,112 @@
...
+  it('It should fail if subject of a mail is not specified', () => {
+    mailTestNoSubject.to('sample@mail.com')
+      .message(`
+        Hi ${data.name},
+        Verify via this link ${data.link}
+      `)
+      .send()
+      .then(() => {
+      })
+      .catch((err) => {
```
Comment (bahdcoder) [inline test/index.spec.js]: This test is evergreen. It never fails. I suggest changing the test to this:

```js
return mailTestNoSubject.to('sample@mail.com')
      .message(`
        Hi ${data.name},
        Verify via this link ${data.link}
      `)
      .send()
      .catch((err) => {
        expect(err.message.toString()).to.be.equal('Bad Request');
        expect(err.code.toString()).to.be.equal('400');
      });
```

That's the effective way of testing promises. You need to return the promise so that the test running waits till it's completed, then runs assertions. At the moment, the assertions are not even run.

File: test/index.spec.js [hunk 3]
```diff
@@ -1,13 +1,112 @@
...
+  });
+  it('It should be able to send mail to multiple recipients', () => {
+    mailTest.to(['john@mail.com', 'jane@mail.com', 'tony@mail.com'])
+      .message(`
+        Hi ${data.name},
+        Verify via this link ${data.link}
+      `)
+      .send()
+      .then((res) => {
+        expect(res.statusCode).to.have.statusCode(202);
```
Comment (bahdcoder) [inline test/index.spec.js]: This is overgreen too. Experiment by changing `202` to `550` and watch the test still pass. lol. Same recommendation as above. Return the promise.

```js
return mailTest.to(['john@mail.com', 'jane@mail.com', 'tony@mail.com'])
      .message(`
        Hi ${data.name},
        Verify via this link ${data.link}
      `)
      .send()
      .then((res) => {
        expect(res.statusCode).to.have.statusCode(202);
      })
```

File: test/index.spec.js [hunk 4]
```diff
@@ -1,13 +1,112 @@
...
+      })
+      .catch(() => {
+      });
+  });
+  it('An instance of the Mail class should be an object', () => {
```
Comment (bahdcoder) [inline test/index.spec.js]: Every instance of a class is an object. No need for this test really.

Comment (devPinheiro) [reply test/index.spec.js]: I have removed this test.

File: test/index.spec.js (hunk 3 above)
Comment (devPinheiro) [reply test/index.spec.js]: fixed

File: .env.sample (hunk 1 above)
Comment (devPinheiro) [reply .env.sample]: readme.md file updated and I have shared the API key credential with team members on slack

File: test/index.spec.js [hunk 5]
```diff
@@ -1,13 +1,123 @@
...
+chai.should();
 
 const add = 2 + 2;
-describe('Sample Test', () => {
-    it('Add number', () => {
-       expect(add).to.eql(4)
-    })
-    
-})
+describe("Sample Test", () => {
```
Comment (bahdcoder) [inline test/index.spec.js]: Let's get rid of this test. Also, let's reorganise this test to match the project: `test/helpers/mail.spec.js`.

File: test/index.spec.js [hunk 6]
```diff
@@ -1,13 +1,123 @@
...
+  it("Add number", () => {
+    expect(add).to.eql(4);
+  });
+});
+
+// tests for helper Mail class
+describe("Tests for helper Mail class", () => {
```
Comment (bahdcoder) [inline test/index.spec.js]: Let the description be `The Mail class`

File: test/index.spec.js [hunk 7]
```diff
@@ -1,13 +1,123 @@
...
+    // create a new instance of the Mail object without a subject
+    mailTestNoSubject = new Mail();
+  });
+  // sample data to use as body of the  mail
+  const data = {
+    name: "John",
+    link: "somelinktoverifyuser"
+  };
+
+  it("It should fail if no recipient is specified", () => {
```
Comment (bahdcoder) [inline test/index.spec.js]: Remove the `It` from all test descriptions. It's very important that the tests be readable as they provide low level documentation for the project

File: test/index.spec.js (hunk 5 above)
Comment (devPinheiro) [reply test/index.spec.js]: I resolved it.

File: test/index.spec.js (hunk 6 above)
Comment (devPinheiro) [reply test/index.spec.js]: fixed

File: test/index.spec.js (hunk 7 above)
Comment (devPinheiro) [reply test/index.spec.js]: I have fixed this.
